File: sdk-node/src/core/telemetry.ts
import { API_VERSION } from '@/api-version'
import { SDK_VERSION } from '@/version'

/**
 * Telemetry header builder + last-request tracker.
 *
 * The header name `Threecommon-Client-Telemetry`
 * Disable globally with `telemetry: false` on client config or at runtime via
 *`client.disableTelemetry()`.
 *
 * @internal
 */

/**
 * Snapshot of the previous request, attached to the next request as a header.
 *
 * @internal
 */
export interface LastRequestMetric {
  readonly method: string
  readonly path: string
  readonly status: number | undefined
  readonly durationMs: number
  readonly requestId: string | undefined
}

/**
 * Stateful telemetry tracker. One instance per {@link ThreeCommon} client.
 *
 * @internal
 */
export class Telemetry {
  private enabledFlag: boolean
  private last: LastRequestMetric | undefined

  public constructor(enabled: boolean) {
    this.enabledFlag = enabled
    this.last = undefined
  }

  public isEnabled(): boolean {
    return this.enabledFlag
  }

  public disable(): void {
    this.enabledFlag = false
    this.last = undefined
  }

  /** Record a completed request. No-op when disabled. */
  public record(metric: LastRequestMetric): void {
    if (!this.enabledFlag) return
    this.last = metric
  }

  /**
   * Build the value for the next request's `Threecommon-Client-Telemetry`
   * header, or return `undefined` if telemetry is disabled.
   */
  public buildHeaderValue(): string | undefined {
    if (!this.enabledFlag) return undefined

    const payload = {
      lang: 'node',
      sdk: SDK_VERSION,
      api: API_VERSION,
      last:
        this.last === undefined
          ? undefined
          : {
              m: this.last.method,
              p: this.last.path,
              s: this.last.status,
              d: this.last.durationMs,
            },
    }

    return JSON.stringify(payload)
  }
}

File: sdk-python/src/threecommon/_core/telemetry.py
"""Threecommon-Client-Telemetry header builder + last-request tracker.

The header carries SDK version, language, and the previous request's
latency snapshot.

both sync and async paths share one [Telemetry][threecommon._core.telemetry.Telemetry]
instance per client.
"""

from __future__ import annotations

import json
import threading
from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class _Snapshot:
    method: str
    path: str
    status: int
    duration_ms: int


class Telemetry:
    """Tracks one previous-request snapshot and emits the header value."""

    __slots__ = ("_enabled", "_last", "_lock")

    def __init__(self, *, enabled: bool) -> None:
        self._enabled = enabled
        self._last: _Snapshot | None = None
        self._lock = threading.Lock()

    @property
    def enabled(self) -> bool:
        with self._lock:
            return self._enabled

    def disable(self) -> None:
        """Turn telemetry off and clear the cached snapshot."""
        with self._lock:
            self._enabled = False
            self._last = None

    def record(self, *, method: str, path: str, status: int, duration_seconds: float) -> None:
        """Store a snapshot of the just-completed request. No-op when disabled."""
        with self._lock:
            if not self._enabled:
                return
            self._last = _Snapshot(
                method=method,
                path=path,
                status=status,
                duration_ms=int(duration_seconds * 1000),
            )

    def header_value(self, *, sdk_version: str, api_version: str) -> str | None:
        """Return the JSON header value, or ``None`` to omit the header."""
        with self._lock:
            if not self._enabled:
                return None
            last = self._last

        payload: dict[str, object] = {
            "lang": "python",
            "sdk": sdk_version,
            "api": api_version,
        }
        if last is not None:
            payload["last"] = {
                "m": last.method,
                "p": last.path,
                "s": last.status,
                "d": last.duration_ms,
            }
        return json.dumps(payload, separators=(",", ":"))
